src/react/hooks/useFragment.ts

```diff
@@ -65,38 +65,33 @@ function _useFragment<TData = any, TVars = OperationVariables>(
 ): UseFragmentResult<TData> {
   const { cache } = useApolloClient(options.client);
 
-  const diffOptions = useDeepMemo<Cache.DiffOptions<TData, TVars>>(() => {
+  const resultRef = React.useRef<UseFragmentResult<TData>>();
+  const stableOptions = useDeepMemo(() => options, [options]);
+
+  // Since .next is async, we need to make sure that we
+  // get the correct diff on the next render given new diffOptions
+  React.useMemo(() => {
     const {
       fragment,
       fragmentName,
       from,
       optimistic = true,
       ...rest
-    } = options;
-
-    return {
-      ...rest,
-      returnPartialData: true,
-      id: typeof from === "string" ? from : cache.identify(from),
-      query: cache["getFragmentDoc"](fragment, fragmentName),
-      optimistic,
-    };
-  }, [options]);
+    } = stableOptions;
 
-  const resultRef = useLazyRef<UseFragmentResult<TData>>(() =>
-    diffToResult(cache.diff<TData>(diffOptions))
-  );
-
-  const stableOptions = useDeepMemo(() => options, [options]);
-
-  // Since .next is async, we need to make sure that we
-  // get the correct diff on the next render given new diffOptions
-  React.useMemo(() => {
-    resultRef.current = diffToResult(cache.diff<TData>(diffOptions));
-  }, [diffOptions, cache]);
+    resultRef.current = diffToResult(
+      cache.diff<TData>({
+        ...rest,
+        returnPartialData: true,
+        id: typeof from === "string" ? from : cache.identify(from),
+        query: cache["getFragmentDoc"](fragment, fragmentName),
+        optimistic,
+      })
+    );
+  }, [stableOptions, cache]);
 
   // Used for both getSnapshot and getServerSnapshot
-  const getSnapshot = React.useCallback(() => resultRef.current, []);
+  const getSnapshot = React.useCallback(() => resultRef.current!, []);
 
   return useSyncExternalStore(
     React.useCallback(
```

Comment:
What do you think about this?
```suggestion
  // Since .next is async, we need to make sure that we
  // get the correct diff on the next render given new diffOptions
  const currentDiff = React.useMemo(() => {
    const {
      fragment,
      fragmentName,
      from,
      optimistic = true,
      ...rest
    } = stableOptions;

    return diffToResult(
      cache.diff<TData>({
        ...rest,
        returnPartialData: true,
        id: typeof from === "string" ? from : cache.identify(from),
        query: cache["getFragmentDoc"](fragment, fragmentName),
        optimistic,
      })
    );
  }, [stableOptions, cache]);

  // Used for both getSnapshot and getServerSnapshot
  const getSnapshot = React.useCallback(
    () => resultRef.current || currentDiff,
    [currentDiff]
  );

  return useSyncExternalStore(
    React.useCallback(
```

and then in the cleanup 

```js
        return () => {
          resultRef.current = undefined;
          subscription.unsubscribe();
          clearTimeout(lastTimeout);
        };
```

Reply:
Looks like that causes some tests failures. I'll dig in a bit more after planning to try and understand why this change causes it because I like this change a lot.

Reply:
Ok seems this fails so many tests because the check to determine if the current result matches the last one fails on the initial render:

```ts
if (equal(result, resultRef.current)) return;
```

With this change in place, `resultRef.current` is `undefined`, so this check will always fail, therefore we get an extra render since the hook thinks the result is different.

This check is really only used for the first result emitted from `watchFragment`, specifically for this case of initialization since `watchFragment` emits its initial  value asynchronously and we want to avoid an extra render for the initial value.

Let me see if comparing against `currentResult` in uSES (which means adding it to the dependency array) causes any failures here.